File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Crc32Ieee.java
package com.freight.hash;

/** crc32_ieee over raw bytes. */
public final class Crc32Ieee implements HashAlgorithm {

    @Override
    public String name() {
        return "crc32_ieee";
    }

    @Override
    public long apply(byte[] data) {
        long crc = 0xFFFFFFFFL;
        for (int i = 0; i < data.length; i++) {
            crc ^= (data[i] & 0xFF);
            for (int bit = 0; bit < 8; bit++) {
                crc = ((crc & 1L) != 0L) ? ((crc >>> 1) ^ 0xEDB88320L) : (crc >>> 1);
            }
        }
        return (crc ^ 0xFFFFFFFFL) & 0xFFFFFFFFL;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/hashx/crc32_ieee.go
package hashx

// Crc32Ieee computes the crc32_ieee checksum over raw bytes.
func Crc32Ieee(data []byte) uint64 {
	crc := uint32(0xFFFFFFFF)
	for _, b := range data {
		crc ^= uint32(b)
		for bit := 0; bit < 8; bit++ {
			if crc&1 != 0 {
				crc = (crc >> 1) ^ 0xEDB88320
			} else {
				crc >>= 1
			}
		}
	}
	return uint64(crc ^ 0xFFFFFFFF)
}
